### PR TITLE
[App] Fixing Sigterm Handler causing thread lock which caused KeyboardInterrupt to hang

### PR DESCRIPTION
**Pull request by hhsecond**
## What does this PR do?

Multiprocessing runtime was configured to call `terminate` on the work process twice

```python
        except KeyboardInterrupt:
            self.terminate()
            raise
        finally:
            self.terminate()
```

The second `terminate` would cancel the first sigterm handler call but cancellation of a handler wouldn't release the thread lock causing the second sigterm handler to be on a deadlock

Note: We still have issues with uvloop overriding the signal handlers. But that's a different issue

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
Amazing catch !

File: src/lightning_app/runners/multiprocess.py [hunk 1]
```diff
@@ -111,9 +112,11 @@ def dispatch(self, *args: Any, on_before_run: Optional[Callable] = None, **kwarg
             self.app._run()
         except KeyboardInterrupt:
             self.terminate()
+            self._has_triggered_termination = True
             raise
         finally:
-            self.terminate()
+            if not self._has_triggered_termination:
```
[the remark this answers is not in the document]
Comment (hhsecond) [reply src/lightning_app/runners/multiprocess.py:118]: no `self.terminate()` is whats triggering the termination. If it's already triggered, we don't need to retrigger

**Comment by lantiga** [comment]
Late to the party, but great catch @hhsecond!

**Comment by lantiga** [comment]
Late to the party, but great catch @hhsecond!

**Pull request merged**